KAFKA-14584: Deprecate StateChangeLogMerger tool

fvaleri opened the pull request:
This also includes a new test.

fvaleri commented:
@mimaison @OmniaGM @tombentley this is ready for review if any of you have time. Thanks.

OmniaGM commented on server-common/src/main/java/org/apache/kafka/server/util/ToolsUtils.java:
```diff
@@ -100,4 +104,17 @@ public static void prettyPrintTable(
+    /**
+     * Returns a set of duplicated items.
+     */
+    public static <T> Set<T> findDuplicates(Collection<T> collection) {
```
Are we planning to drop `CoreUtils.duplicates` in favor of `ToolsUtils.findDuplicates` at some point as `CoreUtils.duplicates` are only used by cli tools at the moment

fvaleri replied:
That was my idea. I think we can drop CoreUtils.duplicates once the last command using it is migrated. I renamed from ToolsUtils.findDuplicates to ToolsUtils.duplicates, so the intention is clear. WDYT?

OmniaGM replied:
sounds like a plan. Am working on other cli commands that use the `CoreUtils.duplicates` so keep a note to remove it with the last command.

fvaleri replied:
We can create a KAFKA-14525 subtask where we list all pending cleanups/refactorings. You could even reference it in a comment on your PR.

OmniaGM commented on tools/src/main/java/org/apache/kafka/tools/StateChangeLogMerger.java:
```diff
@@ -0,0 +1,329 @@
+    private static Date startDate;
+    private static Date endDate;
+
+    public static void main(String[] args) {
```
It may be better to unify the cli tool classes to use same pattern as `ClusterTools` and `MetadataQuorumCommand`to have `main`, `mainNoExit`, and `execute`. Example is here https://github.com/apache/kafka/blob/72cfc994f5675be349d4494ece3528efed290651/tools/src/main/java/org/apache/kafka/tools/MetadataQuorumCommand.java#L49

fvaleri replied:
Yeah, I thought about that, but there are shared util methods that call Exit.exit() directly, so we should change that first. I think we can address this in a separate PR once most commands are migrated. WDYT?

OmniaGM commented on tools/src/main/java/org/apache/kafka/tools/StateChangeLogMerger.java:
```diff
@@ -0,0 +1,329 @@
+                return;
+            }
+
+            if ((!options.hasFiles() && !options.hasRegex()) || (options.hasFiles() && options.hasRegex())) {
```
I know the original code use this pattern, however, can't we use `CommandLineUtils.checkInvalidArgs` here instead?

OmniaGM commented on tools/src/main/java/org/apache/kafka/tools/StateChangeLogMerger.java:
```diff
@@ -0,0 +1,329 @@
+                    String.format("Provide arguments to exactly one of the two options \"%s\" or \"%s\"", options.filesOpt, options.regexOpt));
+                return;
+            }
+            if (options.hasPartitions() && !options.hasTopic()) {
```
Similiar to the previous comment can't we use `CommandLineUtils.checkRequiredArgs`.

OmniaGM reviewed (commented):
Thanks @fvaleri. Great work, just I left few comments.

fvaleri commented:
Hi Omnia, thanks for the review.

I've addressed some of your comments and answered the others.

clolov commented on build.gradle:
```diff
@@ -1763,6 +1763,7 @@ project(':tools') {
     implementation libs.jacksonJDK8Datatypes
     implementation libs.slf4jApi
     implementation libs.log4j
+    implementation libs.joptSimple
```
I believe we should stop adding this to pull requests 😄. It should have been merged in https://github.com/apache/kafka/pull/13136/files if not before.

mimaison replied:
Yes this PR needs to be rebased on trunk to remove the extra changes like this and `ToolsTestUtils`/`TransactionCommandTest`.

mimaison commented on tools/src/main/java/org/apache/kafka/tools/StateChangeLogMerger.java:
```diff
@@ -0,0 +1,320 @@
+    public static void main(String[] args) {
+        try {
+            StateChangeLogMergerOptions options = new StateChangeLogMergerOptions(args);
+            if (printHelpOrVersion(options) || !validInput(options)) return;
```
What about doing:
```
CommandLineUtils.maybePrintHelpOrVersion(options,
                "A tool for merging the log files from several brokers to reconstruct a unified history of what happened.");
validInput(options);
```
And remove the `printHelpOrVersion()` method and have `validInput()` return `void`?

mimaison replied:
And we could even put `CommandLineUtils.maybePrintHelpOrVersion()` in `validInput()`.

fvaleri replied:
Yes, that was the original intention. Thanks.

clolov reviewed (approved):
Looks great! Once it is rebased I believe this is ready to merge.

mimaison commented on tools/src/test/java/org/apache/kafka/tools/StateChangeLogMergerTest.java:
```diff
@@ -0,0 +1,193 @@
+    @BeforeAll
+    public static void beforeAll() throws IOException {
+        tempFile = Files.createFile(tempDir.resolve("state-change.log"));
+        try (PrintWriter output = new PrintWriter(new FileOutputStream(tempFile.toFile()))) {
```
We can use `Files.newOutputStream(tempFile.toFile().toPath())` instead of `new FileOutputStream`

mimaison reviewed (commented):
While testing the tool, I realized I can't get it to work on any recent state-change log files. I found https://issues.apache.org/jira/browse/KAFKA-7735 which indicates the tool has been broken for a very long time! 
If it's easy to fix, I'm ok doing it, otherwise I wonder if we could delete it as nobody has complained in over 4 years.

fvaleri commented:
@mimaison thanks for spotting this! 

I saw that ticket, but I started to test with an old log file for testing purpose, and then forgot to also test it on a recent log version. I agree that this is not much used, but maybe the logic can be reused to create something similar.

With the latest commit, I make it work with the latest state change log format and addressed your comments. Before merging, we need to rebase with https://github.com/apache/kafka/pull/13195, so that I can also add this tool to the notable changes note.

fvaleri commented:
Given that clearly nobody is using this tool, I'm wondering if we should deprecate in 3.5 and remove in 4.0.

ijuma commented:
Yeah, I think we should consider simply removing the tool if it doesn't work. But if it was exposed in ` bin`, we'd need a KIP for it.

fvaleri commented:
Hi @mimaison, now that KIP-906 is accepted I would say this is ready for the final review. Thanks.

fvaleri committed:
KAFKA-14584: Deprecate StateChangeLogMerger tool

Signed-off-by: Federico Valeri <fedevaleri@gmail.com>

fvaleri commented:
@mimaison I reverted these changes and added the deprecation note as stated in KIP-906 and suggested by @ijuma.

mimaison reviewed (approved):
LGTM, thanks for the PR
